File: benches/fib.py
import time
import csv

def fibonacci(n):
    if n <= 1:
        return n
    else:
        return fibonacci(n-1) + fibonacci(n-2)

if __name__ == "__main__":
    total_time = 0
    with open('fib-py.csv', mode='w') as csv_file:
        fieldnames = ['Run', 'Time']
        writer = csv.DictWriter(csv_file, fieldnames=fieldnames)

        writer.writeheader()
        for i in range(1, 21):
            start_time = time.time()
            fibonacci(45)
            elapsed_time = time.time() - start_time
            total_time += elapsed_time
            print(f"Run {i} Time taken: {elapsed_time} seconds")
            writer.writerow({'Run': i, 'Time': elapsed_time})

        print(f"Average time taken: {total_time/20}s")

File: benches/fib.go
package main

import (
  "encoding/csv"
  "fmt"
  "os"
  "time"
)

func fibonacci(n int) int {
  if n <= 1 {
    return n
  } else {
    return fibonacci(n-1) + fibonacci(n-2)
  }
}

func main() {
  totalTime := time.Duration(0)
  file, err := os.Create("fib-go.csv")
  if err != nil {
    panic(err)
  }
  defer file.Close()

  writer := csv.NewWriter(file)
  defer writer.Flush()

  fmt.Fprintf(file, "Run,Time\n")

  for i := 0; i < 20; i++ {
    startTime := time.Now()
    fibonacci(45)
    endTime := time.Now()
    executionTime := endTime.Sub(startTime).Seconds()
    totalTime += endTime.Sub(startTime)
    fmt.Printf("Run %d time taken: %v seconds\n", i+1, executionTime)
    writer.Flush()
    fmt.Fprintf(file, "%d,%f\n", i+1, executionTime)
  }
  fmt.Printf("Average time taken: %v s\n", totalTime.Seconds()/20)
}
